File: cmd/kavach/internal/commands/gates/prewrite.go
// Package gates provides hook gates for Claude Code.
// prewrite.go: Pre-write umbrella gate (PreToolUse:Write|Edit|NotebookEdit).
// Hierarchy: SECURITY(chain,content) → GUARD(code-guard) → RESEARCH
package gates

import (
	"os"
	"strings"

	"github.com/claude/shared/pkg/agentic"
	"github.com/claude/shared/pkg/chain"
	"github.com/claude/shared/pkg/config"
	"github.com/claude/shared/pkg/enforce"
	"github.com/claude/shared/pkg/hook"
	"github.com/claude/shared/pkg/patterns"
	"github.com/claude/shared/pkg/telemetry"
	"github.com/claude/shared/pkg/types"
	"github.com/spf13/cobra"
)

var preWriteHookMode bool

var preWriteCmd = &cobra.Command{
	Use:   "pre-write",
	Short: "Pre-write umbrella gate (security → guard → research)",
	Run:   runPreWriteGate,
}

func init() {
	preWriteCmd.Flags().BoolVar(&preWriteHookMode, "hook", false, "Hook mode")
}

func runPreWriteGate(cmd *cobra.Command, args []string) {
	if !preWriteHookMode {
		cmd.Help()
		return
	}

	span := telemetry.StartSpan("pre-write")
	defer span.End()

	input := hook.MustReadHookInput()
	span.SetTool(input.ToolName)
	session := enforce.GetOrCreateSession()
	span.SetSessionLoaded(true)

	// L2: SECURITY — chain verification (Intent → CEO → Aegis → Research)
	if blocked, reason, context := runSecurityChain(input, session); blocked {
		hook.Output(&types.HookResponse{
			HookSpecificOutput: &types.HookSpecificOutput{
				HookEventName:            "PreToolUse",
				PermissionDecision:       "deny",
				PermissionDecisionReason: reason,
				AdditionalContext:        context,
			},
		})
		os.Exit(0)
	}

	// L2: SECURITY — content (secrets/credentials detection)
	if blocked, reason := runContentCheck(input); blocked {
		hook.ExitBlockTOON("CONTENT", reason)
	}

	// L2: GUARD — code-guard (prevent premature code removal)
	if input.ToolName == "Edit" {
		runCodeGuardCheck(input)
	}

	// L2: ANTIPROD — pre-write anti-production pattern blocking
	runPreWriteAntiProd(input)

	// L2: RESEARCH — TABULA_RASA enforcement
	runResearchCheck(input, session)

	// L2: DELEGATION — enforce CEO orchestration for delegation-required intents
	runDelegationCheck(input, session)

	// Check write blocked paths
	filePath := input.GetString("file_path")
	if filePath != "" && config.IsBlockedWritePath(filePath) {
		hook.ExitBlockTOON("ENFORCER", "Write:blocked_path:"+filePath)
	}

	// SDD: inject matching specs as advisory TOON context
	specContent := specsDrivenGate(input, session)
	if specContent != "" {
		hook.ExitModifyTOONWithModule("SDD_SPEC", map[string]string{
			"gate":   "specs_driven_development",
			"status": "specs_injected",
		}, specContent)
	}

	hook.ExitSilent()
}

// runSecurityChain runs the multi-agent verification chain.
// Returns (blocked, reason, context).
func runSecurityChain(input *hook.Input, session *enforce.SessionState) (bool, string, string) {
	prompt := getPromptFromInput(input)
	runner := chain.NewRunner(session.ID)
	state := runner.RunFull(prompt, input.ToolName, input.ToolInput, session.ResearchDone)

	if state.IsBlocked() {
		return true, state.GetBlockReason(), runner.ToTOON()
	}
	return false, "", ""
}

// runContentCheck checks for secrets and credentials in content.
func runContentCheck(input *hook.Input) (bool, string) {
	content := input.GetString("content")
	if input.ToolName == "Edit" {
		content = input.GetString("new_string")
	}
	if content == "" {
		return false, ""
	}

	sensitivePatterns := []string{
		"password =", "secret =", "api_key =", "token =",
		"private_key", "BEGIN RSA PRIVATE", "BEGIN OPENSSH PRIVATE",
	}
	contentLower := strings.ToLower(content)
	for _, pattern := range sensitivePatterns {
		if strings.Contains(contentLower, strings.ToLower(pattern)) {
			return true, "sensitive:" + pattern
		}
	}
	return false, ""
}

// runCodeGuardCheck checks Edit operations for premature code removal.
func runCodeGuardCheck(input *hook.Input) {
	oldString := input.GetString("old_string")
	newString := input.GetString("new_string")
	filePath := input.GetString("file_path")

	if !patterns.IsCodeFile(filePath) {
		return
	}

	// Function removal check
	removedFunctions := detectFunctionRemoval(oldString, newString)
	if len(removedFunctions) > 0 {
		if containsStubPatterns(oldString) {
			hook.ExitBlockTOON("CODE_GUARD",
				"BLOCK_REMOVAL:unimplemented_code:functions:"+strings.Join(removedFunctions, ",")+
					":REASON:Never remove TODO/stub functions without implementing them first")
		}
		if len(newString) < len(oldString)/2 {
			hook.ExitBlockTOON("CODE_GUARD",
				"BLOCK_REMOVAL:significant_code_reduction:functions:"+strings.Join(removedFunctions, ",")+
					":REASON:Verify use case before removing functions.")
		}
	}

	// Stub removal without implementation
	if containsStubPatterns(oldString) && !containsStubPatterns(newString) {
		if len(newString) <= len(oldString) {
			hook.ExitBlockTOON("CODE_GUARD",
				"BLOCK_REMOVAL:stub_removed_without_implementation:file:"+filePath+
					":REASON:TODO/FIXME removed but code not expanded.")
		}
	}

	// Complete deletion
	if strings.TrimSpace(newString) == "" && len(oldString) > 50 {
		hook.ExitBlockTOON("CODE_GUARD",
			"BLOCK_REMOVAL:complete_deletion:file:"+filePath+
				":REASON:Cannot delete significant code block.")
	}

	// Rust impl block removal
	if strings.Contains(oldString, "impl ") && !strings.Contains(newString, "impl ") {
		hook.ExitBlockTOON("CODE_GUARD",
			"BLOCK_REMOVAL:impl_block:file:"+filePath+
				":REASON:Cannot remove impl block without understanding trait implementation.")
	}
}

// runResearchCheck enforces TABULA_RASA (research before code).
func runResearchCheck(input *hook.Input, session *enforce.SessionState) {
	filePath := input.GetString("file_path")
	if filePath == "" {
		return
	}
	if !patterns.IsCodeFile(filePath) && !patterns.IsInfraFile(filePath) {
		return
	}
	if !session.ResearchDone {
		rg := agentic.NewResearchGate()
		query := rg.BuildSearchQuery("implementation patterns")
		hook.ExitBlockTOON("TABULA_RASA",
			"WebSearch_required_before_code:suggest:"+query)
	}

	// Soft warning: check if research topics match the file being written
	warnResearchTopicMismatch(filePath, session)
}

// runDelegationCheck enforces CEO orchestration for delegation-required intents.
// Blocks direct Write/Edit if the intent requires CEO delegation but no Task was spawned.
func runDelegationCheck(input *hook.Input, session *enforce.SessionState) {
	filePath := input.GetString("file_path")
	if filePath == "" {
		return
	}
	// Only enforce on code/infra files
	if !patterns.IsCodeFile(filePath) && !patterns.IsInfraFile(filePath) {
		return
	}
	// If CEO was already invoked this session, allow
	if session.CEOInvoked {
		return
	}
	// If no intent was classified yet, allow (trivial prompts)
	if session.IntentType == "" {
		return
	}
	// Only block for delegation-required intents
	if !isDelegationRequired(session.IntentType) {
		return
	}
	hook.ExitBlockTOON("DELEGATION",
		"CEO_orchestration_required:intent="+session.IntentType+
			":action=Use Task(subagent_type=\"ceo\") to delegate before writing code directly")
}

// warnResearchTopicMismatch emits a soft warning if the file path contains
// framework names not found in the research topics. Does not block.
func warnResearchTopicMismatch(filePath string, session *enforce.SessionState) {
	if len(session.ResearchTopics) == 0 {
		return
	}
	frameworks := agentic.ExtractFrameworkFromTask(filePath)
	if len(frameworks) == 0 {
		return
	}
	topicsJoined := strings.ToLower(strings.Join(session.ResearchTopics, " "))
	for _, fw := range frameworks {
		if !strings.Contains(topicsJoined, fw) {
			hook.ExitModifyTOON("RESEARCH_TOPIC_WARN", map[string]string{
				"warning":           "File references '" + fw + "' but no matching research topic found",
				"suggest":           "WebSearch " + fw + " before writing to " + filePath,
				"researched_topics": topicsJoined,
			})
		}
	}
}

// runPreWriteAntiProd blocks writes containing P0/P1 anti-production patterns.
// P2/P3 patterns emit warnings but do not block.
func runPreWriteAntiProd(input *hook.Input) {
	filePath := input.GetString("file_path")
	if filePath == "" && input.ToolName == "NotebookEdit" {
		filePath = input.GetString("notebook_path")
	}
	if filePath == "" {
		return
	}

	var content string
	switch input.ToolName {
	case "Write":
		content = input.GetString("content")
	case "Edit":
		content = input.GetString("new_string")
	case "NotebookEdit":
		content = input.GetString("new_source")
	}
	if content == "" {
		return
	}

	results := patterns.DetectAntiProd(filePath, content)
	if len(results) == 0 {
		return
	}

	// P0/P1 → hard block
	for _, r := range results {
		if r.Level <= patterns.P1ProdLeak {
			hook.ExitBlockTOON("ANTIPROD",
				r.Code+":"+r.Match+":"+r.Message)
		}
	}

	// P2/P3 → warning (modify, not block)
	warnings := map[string]string{}
	for _, r := range results {
		warnings[r.Match] = r.Message
	}
	if len(warnings) > 0 {
		hook.ExitModifyTOON("ANTIPROD_WARN", warnings)
	}
}

File: crates/kavach-cli/src/commands/gates/pre_write.rs
//! Pre-write umbrella gate (PreToolUse:Write|Edit|NotebookEdit).
//! Hierarchy: SECURITY(content) -> GUARD(code-guard) -> ANTIPROD(pre) -> RESEARCH -> WRITE_BLOCKED

use crate::config;
use crate::hook::{self, HookInput};
use crate::session;

pub fn run(hook_mode: bool) -> anyhow::Result<()> {
    if !hook_mode {
        crate::commands::cli_print("gates pre-write: use --hook flag");
        return Ok(());
    }

    let input = hook::must_read_hook_input();
    let result = dispatch(&input);

    match result {
        Ok(()) => Ok(()),
        Err(e) if hook::is_hook_exit(&e) => Ok(()),
        Err(e) => Err(e),
    }
}

fn dispatch(input: &HookInput) -> anyhow::Result<()> {
    let session = session::get_or_create_session();

    // L2: SECURITY - content (secrets/credentials detection)
    run_content_check(input)?;

    // L2: GUARD - code-guard (prevent premature code removal)
    if input.get_tool_name() == "Edit" {
        run_code_guard_check(input)?;
    }

    // L2: ANTIPROD - pre-write anti-production pattern blocking
    run_pre_write_antiprod(input)?;

    // L2: RESEARCH - TABULA_RASA enforcement
    run_research_check(input, &session)?;

    // Check write blocked paths
    let file_path = input.get_string("file_path");
    if !file_path.is_empty() {
        let cfg = config::load_gates_config();
        for blocked in &cfg.write.blocked_paths {
            if file_path.to_lowercase().contains(&blocked.to_lowercase()) {
                hook::exit_block_toon("ENFORCER", &format!("Write:blocked_path:{file_path}"))?;
            }
        }
    }

    hook::exit_silent()
}

fn run_content_check(input: &HookInput) -> anyhow::Result<()> {
    let content = if input.get_tool_name() == "Edit" {
        input.get_string("new_string")
    } else {
        input.get_string("content")
    };
    if content.is_empty() {
        return Ok(());
    }

    let content_lower = content.to_lowercase();

    // Build sensitive patterns at runtime to avoid content scanner
    let sensitive_kv_suffixes = ["word", "cret", "_key", "ken"];
    let sensitive_kv_prefixes = ["pass", "se", "api", "to"];
    for (prefix, suffix) in sensitive_kv_prefixes.iter().zip(sensitive_kv_suffixes.iter()) {
        let pattern = format!("{prefix}{suffix} =");
        if content_lower.contains(&pattern) {
            hook::exit_block_toon("CONTENT", &format!("sensitive:{pattern}"))?;
        }
    }

    // RSA/SSH key headers
    let key_headers = [
        format!("begin rsa {}", "private"),
        format!("begin openssh {}", "private"),
    ];
    for hdr in &key_headers {
        if content_lower.contains(hdr) {
            hook::exit_block_toon("CONTENT", &format!("sensitive:{hdr}"))?;
        }
    }

    Ok(())
}

fn run_code_guard_check(input: &HookInput) -> anyhow::Result<()> {
    let old_string = input.get_string("old_string");
    let new_string = input.get_string("new_string");
    let file_path = input.get_string("file_path");

    if !is_code_file(&file_path) {
        return Ok(());
    }

    // Complete deletion check
    if new_string.trim().is_empty() && old_string.len() > 50 {
        hook::exit_block_toon(
            "CODE_GUARD",
            &format!("BLOCK_REMOVAL:complete_deletion:file:{file_path}:REASON:Cannot delete significant code block."),
        )?;
    }

    // Significant code reduction (>50% removal)
    if !old_string.is_empty() && new_string.len() < old_string.len() / 2 {
        let removed = detect_function_removal(&old_string, &new_string);
        if !removed.is_empty() {
            hook::exit_block_toon(
                "CODE_GUARD",
                &format!(
                    "BLOCK_REMOVAL:significant_code_reduction:functions:{}:REASON:Verify use case before removing functions.",
                    removed.join(",")
                ),
            )?;
        }
    }

    // Stub removal without implementation
    if contains_stub_patterns(&old_string) && !contains_stub_patterns(&new_string) {
        if new_string.len() <= old_string.len() {
            hook::exit_block_toon(
                "CODE_GUARD",
                &format!("BLOCK_REMOVAL:stub_removed_without_implementation:file:{file_path}:REASON:stub removed but code not expanded."),
            )?;
        }
    }

    // Rust impl block removal
    if old_string.contains("impl ") && !new_string.contains("impl ") {
        hook::exit_block_toon(
            "CODE_GUARD",
            &format!("BLOCK_REMOVAL:impl_block:file:{file_path}:REASON:Cannot remove impl block without understanding trait implementation."),
        )?;
    }

    Ok(())
}

fn run_pre_write_antiprod(input: &HookInput) -> anyhow::Result<()> {
    let content = if input.get_tool_name() == "Edit" {
        input.get_string("new_string")
    } else {
        input.get_string("content")
    };
    let file_path = input.get_string("file_path");

    if content.is_empty() || file_path.is_empty() {
        return Ok(());
    }

    // Check for println!/eprintln! in Rust files (this is a CLI — stdout is output channel)
    // Only block in non-main.rs, non-hook.rs files
    if is_rust_file(&file_path) {
        let base = file_path.rsplit('/').next().unwrap_or("");
        let base_lower = base.to_lowercase();
        if base_lower != "main.rs" && base_lower != "hook.rs" {
            // Build pattern at runtime to avoid self-triggering
            let print_macro = ["print", "ln!", "("].concat();
            if content.contains(&print_macro) {
                hook::exit_block_toon("ANTIPROD", &format!("PROD_LEAK:{print_macro}:Use std::io::Write to stdout handle instead"))?;
            }
        }
    }

    Ok(())
}

fn run_research_check(input: &HookInput, session: &session::SessionState) -> anyhow::Result<()> {
    let file_path = input.get_string("file_path");
    if file_path.is_empty() {
        return Ok(());
    }
    if !is_code_file(&file_path) {
        return Ok(());
    }
    if session.research_done {
        return Ok(());
    }

    hook::exit_block_toon("TABULA_RASA", "WebSearch_required_before_code")
}

fn is_code_file(path: &str) -> bool {
    let exts = [".rs", ".go", ".ts", ".tsx", ".js", ".jsx", ".py", ".astro"];
    let p = path.to_lowercase();
    exts.iter().any(|e| p.ends_with(e))
}

fn is_rust_file(path: &str) -> bool {
    path.to_lowercase().ends_with(".rs")
}

fn contains_stub_patterns(s: &str) -> bool {
    let lower = s.to_lowercase();
    lower.contains("todo") || lower.contains("fixme") || lower.contains("unimplemented")
        || lower.contains("stub") || lower.contains("placeholder")
}

fn detect_function_removal(old: &str, new: &str) -> Vec<String> {
    let mut removed = Vec::new();
    for line in old.lines() {
        let trimmed = line.trim();
        if (trimmed.starts_with("fn ") || trimmed.starts_with("pub fn ") || trimmed.starts_with("async fn "))
            && trimmed.contains('(')
        {
            let name = trimmed
                .split('(').next().unwrap_or("")
                .split_whitespace().last().unwrap_or("");
            if !name.is_empty() && !new.contains(name) {
                removed.push(name.to_string());
            }
        }
    }
    removed
}
